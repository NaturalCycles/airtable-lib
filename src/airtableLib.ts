import { _Memo } from '@naturalcycles/js-lib'
import type { AirtableApi } from './airtable.api.js'
import type { AirtableLibCfg } from './airtable.model.js'

export class AirtableLib {
  constructor(public airtableServiceCfg: AirtableLibCfg) {}

  @_Memo()
  async api(): Promise<AirtableApi> {
    // lazy-loading the library
    const airtableApi = (await import('airtable')).default as any as AirtableApi

    const { apiKey } = this.airtableServiceCfg

    airtableApi.configure({
      // endpointURL: 'https://api.airtable.com',
      apiKey,
      // Default is 5 minutes, we override the default to 40 seconds
      requestTimeout: 40_000,
    })
    return airtableApi
  }
}
