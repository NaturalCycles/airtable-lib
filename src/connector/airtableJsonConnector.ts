import { fs2 } from '@naturalcycles/nodejs-lib'
import type {
  AirtableBaseDaoCfg,
  AirtableConnector,
  AirtableDaoOptions,
} from '../airtable.model.js'

export const AIRTABLE_CONNECTOR_JSON = Symbol('AIRTABLE_CONNECTOR_JSON')

export interface AirtableJsonConnectorCfg {
  /**
   * Directory where json cache is stored.
   * Will be stored as `${cacheDir}/${baseName}.json`
   */
  cacheDir: string
}

export class AirtableJsonConnector<BASE = any> implements AirtableConnector<BASE> {
  constructor(private cfg: AirtableJsonConnectorCfg) {}

  readonly TYPE = AIRTABLE_CONNECTOR_JSON

  async fetch(baseDaoCfg: AirtableBaseDaoCfg<BASE>, _opt: AirtableDaoOptions = {}): Promise<BASE> {
    // require ensures the read operation is cached
    // return require(this.jsonPath)
    // NO: going in favor of async interface for all connectors

    return await fs2.readJsonAsync(`${this.cfg.cacheDir}/${baseDaoCfg.baseName}.json`)
  }

  async upload(base: BASE, baseDaoCfg: AirtableBaseDaoCfg<BASE>): Promise<void> {
    const jsonPath = `${this.cfg.cacheDir}/${baseDaoCfg.baseName}.json`
    await fs2.outputJsonAsync(jsonPath, base, { spaces: 2 })
  }
}
