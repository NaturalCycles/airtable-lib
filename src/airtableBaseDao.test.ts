import { MOCK_TS_2018_06_21, mockTime } from '@naturalcycles/dev-lib/dist/testing/index.js'
import { beforeEach, expect, test } from 'vitest'
import { AirtableLib } from './airtableLib.js'
import { AIRTABLE_CONNECTOR_JSON } from './connector/airtableJsonConnector.js'
import { mockBaseDao } from './test/airtable.mock.js'

beforeEach(() => {
  mockTime()
})

const airtableLib = new AirtableLib({
  apiKey: 'fakekey',
})

test('getCache', async () => {
  const baseDao = mockBaseDao(await airtableLib.api(), 'baseId')

  expect(baseDao.lastChanged).toBeUndefined()
  expect(baseDao.lastFetchedMap.get(AIRTABLE_CONNECTOR_JSON)).toBeUndefined()

  const cache = await baseDao.getCache()
  // console.log(cache)

  expect(cache).toMatchSnapshot()

  expect(baseDao.lastChanged).toBeUndefined() // accessing the cache and lazy-loading it is not considered as "changed"
})

test('cacheUpdated$', async () => {
  const baseDao = mockBaseDao(await airtableLib.api(), 'baseId')

  let updatedTimes = 0
  baseDao.cacheUpdatedListeners.push(() => updatedTimes++)

  expect(updatedTimes).toBe(0)

  await baseDao.getCache() // should trigger cacheUpdated$
  expect(updatedTimes).toBe(1)

  await baseDao.getCache() // should NOT trigger cacheUpdated$
  expect(updatedTimes).toBe(1)

  const fakeCache: any = { table1: [{ airtableId: 'asd' }] }
  baseDao.setCache(fakeCache) // trigger
  expect(updatedTimes).toBe(2)
  expect(baseDao.lastChanged).toBe(MOCK_TS_2018_06_21)

  baseDao.setCache(fakeCache) // NOT trigger (data is the same)
  expect(updatedTimes).toBe(2)

  baseDao.setCache(undefined) // trigger
  expect(updatedTimes).toBe(3)
})

// todo: getById should lazyLoad
// todo: getTableRecords should lazyLoad
