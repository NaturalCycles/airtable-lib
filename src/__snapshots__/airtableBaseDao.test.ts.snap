// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getCache 1`] = `
{
  "categories": [
    {
      "airtableId": "reczAHtt0JZ4Uq97s",
      "id": "category1",
      "users": [
        "recKeeT3hUQXfly9T",
      ],
    },
    {
      "airtableId": "recGBC7azpVH4Xhfu",
      "id": "category2",
      "users": [
        "recCOtGXUSWzZEbzf",
      ],
    },
    {
      "airtableId": "recy3fSNv0w3yoSBh",
      "id": "category3",
      "users": [
        "recPJPfpwUOpxhd1s",
      ],
    },
  ],
  "permissions": [
    {
      "airtableId": "recKIx5gXMF6sFz9X",
      "descr": "description 1
long text",
      "id": "permission1",
      "parent": [],
      "pub": true,
      "roles": [
        "recFt2k0XC9Lh4MXv",
        "recZuSfp2VFUwzlkZ",
      ],
    },
    {
      "airtableId": "recfA6bfrDSLoLMOR",
      "descr": "description 2
long text",
      "id": "permission2",
      "parent": [
        "recKIx5gXMF6sFz9X",
      ],
      "pub": true,
      "roles": [
        "recXRZ20Y9RRXK1nf",
        "recZuSfp2VFUwzlkZ",
      ],
    },
    {
      "airtableId": "recTXd1ZiyRV6MkCd",
      "descr": "description 3
long text",
      "id": "permission3",
      "parent": [
        "recfA6bfrDSLoLMOR",
      ],
      "pub": true,
      "roles": [
        "recZuSfp2VFUwzlkZ",
      ],
    },
    {
      "airtableId": "recqXONSau2ByYuAP",
      "descr": "description 4
long text",
      "id": "permission4",
      "parent": [],
      "pub": true,
      "roles": [
        "recMtSLHe2fnVOZoF",
      ],
    },
    {
      "airtableId": "recOcMXcycSymXEGn",
      "descr": "description 5
long text",
      "id": "permission5",
      "parent": [
        "recTXd1ZiyRV6MkCd",
      ],
      "pub": true,
      "roles": [
        "recqHMTb4mCFLKumG",
      ],
    },
    {
      "airtableId": "rech4OufSiuAOrc6Q",
      "descr": "description 6
long text",
      "id": "permission6",
      "parent": [],
      "roles": [
        "recqHMTb4mCFLKumG",
      ],
    },
  ],
  "roles": [
    {
      "airtableId": "recZuSfp2VFUwzlkZ",
      "id": "role1",
      "permissions": [
        "recKIx5gXMF6sFz9X",
        "recTXd1ZiyRV6MkCd",
        "recfA6bfrDSLoLMOR",
      ],
      "pub": true,
      "users": [
        "recKeeT3hUQXfly9T",
      ],
    },
    {
      "airtableId": "recMtSLHe2fnVOZoF",
      "id": "role2",
      "permissions": [
        "recqXONSau2ByYuAP",
      ],
      "users": [
        "recCOtGXUSWzZEbzf",
      ],
    },
    {
      "airtableId": "recFt2k0XC9Lh4MXv",
      "id": "role3",
      "permissions": [
        "recKIx5gXMF6sFz9X",
      ],
      "pub": true,
      "users": [
        "recCOtGXUSWzZEbzf",
      ],
    },
    {
      "airtableId": "recqHMTb4mCFLKumG",
      "id": "role4",
      "permissions": [
        "rech4OufSiuAOrc6Q",
        "recOcMXcycSymXEGn",
      ],
      "pub": true,
      "users": [],
    },
    {
      "airtableId": "rec6tdVmScqKRXDkC",
      "id": "role5",
      "permissions": [],
      "users": [
        "rectsieTOdyjEBKGK",
      ],
    },
    {
      "airtableId": "recXRZ20Y9RRXK1nf",
      "id": "role6",
      "permissions": [
        "recfA6bfrDSLoLMOR",
      ],
      "users": [
        "recBwvy2cmRfKXd3G",
      ],
    },
  ],
  "users": [
    {
      "airtableId": "recKeeT3hUQXfly9T",
      "category": [
        "reczAHtt0JZ4Uq97s",
      ],
      "email": "user1@email.com",
      "id": "user1",
      "images": [
        {
          "filename": "oculus_go_product_and_box.jpg",
          "id": "att8BiHM9PcJ1MkSM",
          "size": 18497,
          "thumbnails": {
            "full": {
              "height": 3000,
              "url": "https://dl.airtable.com/.attachmentThumbnails/d3a1ba7b453ae1625185acedf0c092d5/a597d908",
              "width": 3000,
            },
            "large": {
              "height": 360,
              "url": "https://dl.airtable.com/.attachmentThumbnails/5c3bd24479c1ed3c0f317ae4fe4f7a3d/ac170d70",
              "width": 640,
            },
            "small": {
              "height": 36,
              "url": "https://dl.airtable.com/.attachmentThumbnails/cacbc23ff8d66b261a9f572a98a4de78/abf6388d",
              "width": 64,
            },
          },
          "type": "image/jpeg",
          "url": "https://dl.airtable.com/.attachments/822818b7f9dfce155a2decad0183e635/168ac624/oculus_go_product_and_box.jpg",
        },
      ],
      "roles": [
        "recZuSfp2VFUwzlkZ",
      ],
      "tags": [
        "tag1",
      ],
    },
    {
      "airtableId": "recCOtGXUSWzZEbzf",
      "category": [
        "recGBC7azpVH4Xhfu",
      ],
      "email": "user2@email.com",
      "id": "user2",
      "images": [],
      "roles": [
        "recMtSLHe2fnVOZoF",
        "recFt2k0XC9Lh4MXv",
      ],
      "tags": [
        "tag1",
        "tag2",
      ],
    },
    {
      "airtableId": "recPJPfpwUOpxhd1s",
      "category": [
        "recy3fSNv0w3yoSBh",
      ],
      "email": "user3@email.com",
      "id": "user3",
      "images": [],
      "roles": [],
      "tags": [],
    },
    {
      "airtableId": "recBwvy2cmRfKXd3G",
      "category": [],
      "email": "user4@email.com",
      "id": "user4",
      "images": [],
      "roles": [
        "recXRZ20Y9RRXK1nf",
      ],
      "tags": [
        "tag2",
      ],
    },
    {
      "airtableId": "rectsieTOdyjEBKGK",
      "category": [],
      "email": "user5@email.com",
      "id": "user5",
      "images": [],
      "roles": [
        "rec6tdVmScqKRXDkC",
      ],
      "tags": [],
    },
    {
      "airtableId": "recf5MjaK3eslxNy0",
      "category": [],
      "email": "user6@email.com",
      "id": "user6",
      "images": [],
      "roles": [],
      "tags": [],
    },
  ],
}
`;
